### ISSUE TITLE
Provide the ability to use a Custom formatter

### ISSUE THREAD

**Issue by kvpt** [FantasticFiasco/serilog-sinks-http#11]
Hi,

The possibility to select a formatter based on a type is good but to limited.
For a specific usecase I want to a be able to use a custom formatter but with this sink I can't.
The majority of Serilog sinks have this possibility.
Consider to add it.

Thanks.

**Comment by FantasticFiasco** [comment]
Hi kvpt!

I was under the impression that custom formatters where more prone to exist in sink that log text messages, like the console logger, but I can be totally wrong.

Can you provide more information on the problem you are trying to solve, and we'll try to solve that problem together, possibly by implementing new features in this sink?

**Comment by kvpt** [comment]
I want to use this sink to send log to logz.io an ElasticSearch Saas platform.
They provide http(s) endpoints to send log but messages must be specifically formatted.

Each messages must be one line json object, separated be a new line.

Like this :
```json
{"@timestamp": "2017-04-24T13:00", "message": "example 1"}
{"@timestamp": "2017-04-24T13:01", "message": "example 2"}
```

I succeeded to do that, but I had to modify two things :
- Modify the constructor to be able to inject a specific formatter
- Modify the payload format [here](https://github.com/FantasticFiasco/serilog-sinks-http/blob/master/src/Serilog.Sinks.Http/Sinks/Http/Private/Sinks/HttpSink.cs#L93) and [here](https://github.com/FantasticFiasco/serilog-sinks-http/blob/master/src/Serilog.Sinks.Http/Sinks/Http/Private/Http/HttpLogShipper.cs#L206)
For my poc, I have hardcoded it but I think the payload format should be overridable too.

After that : 
- I can inject the Elasticsearch formatter which format json correctly for elasticsearch :  [ElasticsearchJsonFormatter](https://github.com/serilog/serilog-sinks-elasticsearch/blob/master/src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticsearchJsonFormatter.cs)

It's a specific case here, but I think that the case where you must send messages to an external api that you dont own is fairly common. 
For exemple, if you want to upload log in xml format, actually you can't.

I think this sink need a better extensibility, like others sinks :
- Add more configuration parameters to the sink contructor
- Make all the class public to be able to inherit from them

**Comment by FantasticFiasco** [comment]
Thank you for the detailed explanation.

We can definitely sort this out, but before we do I would like the sink to be fully compatible with [Serilog.Settings.Configuration](https://github.com/serilog/serilog-settings-configuration). After that I will start with this.

Would you be able to test a beta version, verifying that I've implemented the feature you are requesting?

**Comment by FantasticFiasco** [comment]
I've implemented support for [Serilog.Settings.Configuration](https://github.com/serilog/serilog-settings-configuration) now, and am ready to start on this issue.

Is this something you still request or are you happy with the solution you have?

**Comment by kvpt** [comment]
The modifications that I've done works good for me, but I think other people may be interested. 
If you want, I can make a pull request to start from that.

**Comment by FantasticFiasco** [comment]
That sounds like a good idea, if you give me push rights to your fork we can collaborate on the PR.

One thing that I would like us to address is the fact that `ITextFormatter` works with single events. In order to produce output with the current format (where the JSON payload has a `events` property) `ITextFormatter` is only half the story. The other half is being able to describe how multiple events are formatted into a HTTP payload.

**Comment by FantasticFiasco** [comment]
We could introduce an interface called `IBatchedTextFormatter` or similar with the following API:

```
public interface IBatchedTextFormatter
{
    void Format(IEnumerable<LogEvent> logEvents, TextWriter output);
}
```

When configuring the HTTP sink, one could specify an implementation of this interface. This would mean that you can write your own implementation that internally use `ElasticsearchJsonFormatter`, i.e. you would be capable of turning the events into a format conforming to logz.io. At the same time there would exist a default implementation in this sink that maintain the existing default JSON format of sent events,

**Comment by kvpt** [comment]
I made the pull request #16 related to this issue.

**Comment by FantasticFiasco** [comment]
Nice!

Can you provide me with push permissions to the fork, and I'll let you squash the commits before we go back to master?

**Comment by FantasticFiasco** [comment]
Ok, I obviously must go to bed since I just now by mistake removed the branch and in consequence the PR. I'll fix it, and I'll also fix the commits that where pushed to your master and not the PR branch.

**Comment by kvpt** [comment]
No problem. 
Don't worry, no need to rush this feature.

**Comment by kvpt** [comment]
Fixed by PR #18.
